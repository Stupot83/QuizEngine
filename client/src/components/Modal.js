import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { createQuiz, updateQuiz, deleteQuiz } from "../actions/actionsForQuizzes";
import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import CloseIcon from "@material-ui/icons/Close";
import AddCircleOutlineIcon from "@material-ui/icons/AddCircleOutline";
import "../sass/Modal.scss";

class Modal extends Component {
    state = {
        quizTitle: "",
        quizCategory: ""
    };

    componentWillReceiveProps(nextProps) {
        if (this.props.edit) {
            this.setState({
                quizTitle: nextProps.title,
                quizCategory: nextProps.category
            });
        }
    }

    onChange = e => {
        this.setState({ [e.target.id]: e.target.value });
    };

    createQuiz = () => {
        let quiz = {
            quizTitle: this.state.quizTitle,
            quizCategory: this.state.quizCategory
        };

        this.props.createQuiz(quiz);
        this.onClose();
    };

    updateQuiz = async id => {
        let quiz = {
            id: this.props.id,
            quizTitle: this.state.quizTitle,
            quizCategory: this.state.quizCategory
        };

        await this.props.updateQuiz(quiz);

        this.onClose();
        window.location.reload();
    };

    deleteQuiz = id => {
        this.props.deleteQuiz(id, this.props.history);
        this.onClose();
    };

    onClose = e => {
        this.props.onClose && this.props.onClose(e);
        this.setState({
            quizTitle: "",
            quizCategory: ""
        });
    };

    render() {
        if (!this.props.modal) {
            return null;
        }

        document.onkeyup = e => {
            if (e.keyCode === 27 && this.props.modal) {
                this.onClose();
            }
        };

        // Edit Quiz Modal
        if (this.props.edit) {
            return (
                <Grid container alignItems="center" justify="center" className="Modal_container">
                    <Grid item xs={12} className="Modal_header_container">
                        <CloseIcon className="Modal_icon" fontSize="large" onClick={this.onClose}></CloseIcon>
                        <Typography variant="h4" className="Modal_header_edit">
                            Edit Quiz
                        </Typography>
                    </Grid>
                    <Grid item xs={12} className="Modal_detail_container">
                        <label>
                            <div id="quizTitleHeader" className="Modal_form_label">
                                Quiz Title
                            </div>
                            <input
                                onChange={this.onChange}
                                value={this.state.quizTitle}
                                id="quizTitle"
                                type="text"
                                placeholder="Quiz Title"
                                className="Modal_form_input"
                            />
                        </label>
                    </Grid>
                    <Grid item xs={12} className="Modal_detail_container">
                        <label>
                            <div id="quizCategoryHeader" className="Modal_form_label">
                                Quiz Category
                            </div>
                            <input
                                onChange={this.onChange}
                                value={this.state.quizCategory}
                                id="quizCategory"
                                type="text"
                                placeholder="Quiz Category"
                                className="Modal_form_input"
                            />
                        </label>
                    </Grid>
                    <Grid item xs={6} className="Modal_form_button_box">
                        <Button
                            id="modalUpdateQuizButton"
                            variant="contained"
                            className="Modal_form_button_edit"
                            size="large"
                            startIcon={<AddCircleOutlineIcon />}
                            onClick={this.updateQuiz.bind(this, this.props.id)}
                        >
                            Update Quiz
                        </Button>
                    </Grid>
                    <Grid item xs={6} className="Modal_form_button_box">
                        <Button
                            id="modalDeleteQuizButton"
                            variant="contained"
                            className="Modal_form_button_delete"
                            size="large"
                            startIcon={<AddCircleOutlineIcon />}
                            onClick={this.deleteQuiz.bind(this, this.props.id)}
                        >
                            Delete Quiz
                        </Button>
                    </Grid>
                </Grid>
            );
        }

        // Create Quiz Modal
        else
            return (
                <Grid container alignItems="center" justify="center" className="Modal_container">
                    <Grid item xs={12} className="Modal_header_container">
                        <CloseIcon className="Modal_icon" fontSize="large" onClick={this.onClose}></CloseIcon>
                        <Typography variant="h4" className="Modal_header_create">
                            Create Quiz
                        </Typography>
                    </Grid>
                    <Grid item xs={12} className="Modal_detail_container">
                        <label>
                            <div id="quizTitleHeader" className="Modal_form_label">
                                Quiz Title
                            </div>
                            <input
                                onChange={this.onChange}
                                value={this.state.quizTitle}
                                id="quizTitle"
                                type="text"
                                placeholder="Quiz Title"
                                className="Modal_form_input"
                            />
                        </label>
                    </Grid>
                    <Grid item xs={12} className="Modal_detail_container">
                        <label>
                            <div id="quizCategoryHeader" className="Modal_form_label">
                                Quiz Category
                            </div>
                            <input
                                onChange={this.onChange}
                                value={this.state.quizCategory}
                                id="quizCategory"
                                type="text"
                                placeholder="Quiz Category"
                                className="Modal_form_input"
                            />
                        </label>
                    </Grid>
                    <Grid item xs={12} className="Modal_form_button_box">
                        <Button
                            id="modalCreateQuizButton"
                            variant="contained"
                            className="Modal_form_button_create"
                            size="large"
                            startIcon={<AddCircleOutlineIcon />}
                            onClick={this.createQuiz}
                        >
                            Create Quiz
                        </Button>
                    </Grid>
                </Grid>
            );
    }
}

const mapStateToProps = state => ({
    auth: state.auth,
    quizzes: state.quizzes
});

export default connect(mapStateToProps, {
    createQuiz,
    updateQuiz,
    deleteQuiz
})(withRouter(Modal));
