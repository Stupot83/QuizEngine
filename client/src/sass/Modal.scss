@keyframes fade-in {
    from {
        opacity: 0;
        transform: translateY(5px);
    }

    to {
        opacity: 1;
        transform: translateY(0px);
    }
}

.Modal_container {
    max-width: 600px !important;
    background: white;
    border: 1px solid #efefef;
    border-radius: 30px;
    animation: fade-in 0.25s;
    transform: scale(1);
    padding: 2rem;
    visibility: visible;
    position: fixed;
    top: 21%;
    height: 500px;
    z-index: 1000;
    box-shadow: 0 1rem 2rem rgba(0, 0, 0, 0.1);

    .Modal_header_container {
        display: block;
        padding-bottom: 10px;
        padding-left: 30px;
        padding-right: 30px;

        .Modal_icon {
            float: right;
            margin-top: 4px;

            &:hover {
                cursor: pointer;
            }
        }

        .Modal_header_create {
            text-align: center;
            color: #34d1bf;
            font-weight: bold;
        }

        .Modal_header_edit {
            text-align: center;
            color: #3454d1;
            font-weight: bold;
        }
    }

    .Modal_detail_container {
        margin-bottom: 1rem;

        .Modal_form_label {
            width: 100%;
            padding-right: 2rem;
            margin-bottom: 0.75rem;
            margin-top: 0.75rem;
            font-size: 14px;
            color: #070707;
            display: block;
        }

        .Modal_form_input {
            background-color: ghostwhite;
            border-bottom: 1px solid #3454d1;
            height: 40px;
            width: 100%;
            line-height: 1.5em;
            outline: 0;
            transition: all 0.25s ease 0s;
            font-size: 16px;
            padding: 0 1rem;

            &:hover {
                transition: all 0.25s ease 0s;
                background-color: #efefef;
            }

            &:focus {
                transition: all 0.25s ease 0s;
                background-color: #efefef;
            }
        }
    }

    .Modal_form_button_box {
        margin-top: 1.75rem;
        text-align: center;

        .Modal_form_button_create {
            min-width: 150px;
            margin-left: 30px;
            margin-right: 30px;
            padding: 15px;
            border-radius: 15px;
            background-color: #34d1bf;
            color: #070707;
            font-weight: bold;
        }

        .Modal_form_button_edit {
            min-width: 150px;
            margin-left: 30px;
            margin-right: 30px;
            padding: 15px;
            border-radius: 15px;
            background-color: #3454d1;
            color: ghostwhite;
            font-weight: bold;
        }

        .Modal_form_button_delete {
            min-width: 150px;
            margin-left: 30px;
            margin-right: 30px;
            padding: 15px;
            border-radius: 15px;
            background-color: #d1345b;
            color: ghostwhite;
            font-weight: bold;
        }
    }
}

.Modal_form_answers_container {
    margin-top: 2.75rem !important;
    height: 175px;
    overflow-y: scroll;
    border-bottom: 1px solid #efefef;

    .Modal_form_answers_details {
        display: flex;
        justify-content: space-between;

        .Modal_form_answers_label {
            width: 100%;
            padding-right: 2rem;
            margin-bottom: 0.75rem;
            margin-top: 0.75rem;
            font-size: 14px;
            color: #070707;
            display: block;

            .Modal_form_answers_input {
                background-color: white;
                border-bottom: 1px solid #3454d1;
                height: 40px;
                width: 100%;
                line-height: 1.5em;
                outline: 0;
                transition: all 0.25s ease 0s;
                font-size: 16px;
                padding: 0 1rem;

                &:hover {
                    transition: all 0.25s ease 0s;
                    background-color: #efefef;
                }

                &:focus {
                    transition: all 0.25s ease 0s;
                    background-color: #efefef;
                }
            }
        }

        .Modal_form_answers_remove {
            align-items: center;
            margin-left: 1rem;
            font-size: 0.5rem;
            margin-top: 0.75rem;
            height: 0.5rem;
            padding: 0.75rem;
            border-color: #d1345b !important;
            border-radius: 30px !important;
            color: #070707 !important;

            &:hover {
                transition: all 0.25s ease 0s;
                background-color: #efefef;
            }

            &:focus {
                transition: all 0.25s ease 0s;
                background-color: #efefef;
            }
        }
    }
}

.Modal_form_answers_button {
    margin-top: 1.75rem;
    min-width: 115px;
    text-align: center;
    border-color: #34d1bf !important;
    border-radius: 30px !important;
    color: #070707 !important;

    &:hover {
        transition: all 0.25s ease 0s;
        background-color: #efefef;
    }

    &:focus {
        transition: all 0.25s ease 0s;
        background-color: #efefef;
    }
}

@media (max-width: 600px) {
    .Modal_container {
        max-width: 480px !important;
        top: 15%;
        height: 625px;
    }
}

.QuestionModal_container {
    max-width: 500px !important;
    background: white;
    border: 1px solid #efefef;
    border-radius: 30px;
    animation: fade-in 0.25s;
    transform: scale(1);
    padding: 2rem;
    visibility: visible;
    position: fixed;
    top: 18%;
    height: 600px;
    z-index: 1000;
    box-shadow: 0 1rem 2rem rgba(0, 0, 0, 0.1);
    overflow-y: scroll;

    .QuestionModal_header_container {
        display: block;
        padding-bottom: 10px;
        padding-left: 30px;
        padding-right: 30px;

        .QuestionModal_icon {
            float: right;
            margin-top: 4px;

            &:hover {
                cursor: pointer;
            }
        }

        .QuestionModal_header_create {
            text-align: center;
            color: #34d1bf;
            font-weight: bold;
        }

        .QuestionModal_header_edit {
            text-align: center;
            color: #3454d1;
            font-weight: bold;
        }
    }

    .Question_form_container {
        width: 100%;

        .QuestionModal_name_container {
            margin-bottom: 1rem;

            .QuestionModal_form_input {
                background-color: white;
                border-bottom: 1px solid #3f51b5;
                height: 40px;
                width: 100%;
                line-height: 1.5em;
                outline: 0;
                transition: all 0.25s ease 0s;
                font-size: 16px;
                padding: 0 1rem;

                &:hover {
                    transition: all 0.25s ease 0s;
                    background-color: #e6e6e6;
                }

                &:focus {
                    transition: all 0.25s ease 0s;
                    background-color: #e6e6e6;
                }
            }
        }

        .QuestionModal_form_label {
            width: 100%;
            padding-right: 2rem;
            margin-bottom: 1.75rem;
            margin-top: 1.75rem;
            font-size: 14px;
            color: #070707;
            display: block;
        }

        .QuestionModal_form_select {
            width: 100%;
            padding-left: 20px;
            outline: none;
        }

        .QuestionModal_button_container {
            display: flex;
            margin-top: 4.75rem;

            .QuestionModal_form_button_edit {
                min-width: 150px;
                margin-left: 30px;
                margin-right: 30px;
                padding: 15px;
                border-radius: 15px;
                background-color: #3454d1;
                color: ghostwhite;
                font-weight: bold;
            }

            .QuestionModal_form_button_delete {
                min-width: 150px;
                margin-left: 30px;
                margin-right: 30px;
                padding: 15px;
                border-radius: 15px;
                background-color: #d1345b;
                color: ghostwhite;
                font-weight: bold;
            }
        }

        .QuestionModal_form_button_box {
            margin-top: 4.75rem;
            text-align: center;

            .QuestionModal_form_create_button {
                min-width: 150px;
                margin-left: 30px;
                margin-right: 30px;
                padding: 15px;
                border-radius: 15px;
                background-color: #34d1bf;
                color: #070707;
                font-weight: bold;
            }
        }
    }
}
