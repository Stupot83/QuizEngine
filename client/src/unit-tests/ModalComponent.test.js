import React from "react";
import { shallow } from "enzyme";
import Modal from "../components/Modal";
import QuizEngineStore from "../quizEngineStore";
import { Provider } from "react-redux";

const mockStore = QuizEngineStore;

describe("Modal component", () => {
    it("should render correctly in debug mode", () => {
        const component = shallow(
            <Provider store={mockStore}>
                <Modal debug />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });

    it("should render correctly", () => {
        const component = shallow(
            <Provider store={mockStore}>
                <Modal />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });
});
