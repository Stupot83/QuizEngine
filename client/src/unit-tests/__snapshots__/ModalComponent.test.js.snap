// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render correctly 1`] = `ShallowWrapper {}`;

exports[`Modal component should render correctly in debug mode 1`] = `ShallowWrapper {}`;
